eclipse-theia/theia
# Fixed confirmExit behaviour for Electron apps

#### What it does


This PR restores the confirmation dialog shown in Electron mode when the user tries to exit without saving changes. On master, there is no confirmation, the app exits without saving.

This relates to an existing Electron issue: https://github.com/electron/electron/issues/9966. Not opening a dialog in the beforeunload handler and using the [will-prevent-unload](https://electronjs.org/docs/api/web-contents#event-will-prevent-unload) event gets around this.

![theia-confirm](https://user-images.githubusercontent.com/12599255/65781960-98da9b80-e144-11e9-8dea-bb63bcc36ca0.png)

#### How to test


Start the Electron app then disable Auto Save. Either:
- Set the "application.confirmExit" preference to "ifRequired", make a change in a file but do not save it.
- Set the "application.confirmExit" preference to "always".

 Then try:
- Opening a new workspace.
- Refreshing the page.
- Closing the app.

In these cases, a confirmation dialog should be displayed.
- Clicking "Yes" should close the modal and unload the window.
- Clicking "No" should close the modal without unloading.

In all other cases, the app should behave as before with no confirmation modal.

#### Review checklist


#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
